File: codeforces/Contests/Round#946/d.cpp
#include <iostream>
#include <unordered_map>
#define endl "\n"
#define ll long long

using namespace std;


void fastIO() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
}
pair<int, int> findMax(string s, int n, unordered_map<char, pair<int, int>> op) {
    int x = 0, y = 0;
    for (int i = 0; i < n; i++) {
        x += op[s[i]].first;
        y += op[s[i]].second;
    }
    return make_pair(x, y);
}

bool checkCloser(int x, int y, int targetX, int targetY, char c) {
    if (c == 'N') {
        if (y < targetY) {
            return true;
        }
    }
    else if (c == 'S') {
        if (y > targetY) {
            return true;
        }
    }
    else if (c == 'E') {
        if (x < targetX) {
            return true;
        }
    }
    else if (c == 'W') {
        if (x > targetX) {
            return true;
        }
    }
    return false;
}

int main() {
    fastIO();

    int t;
    cin >> t;

    while (t--) {
        int n;
        cin >> n;
        string s;
        cin >> s;

        unordered_map<char, pair<int, int>> op;
        op['N'] = make_pair(0, 1);
        op['S'] = make_pair(0, -1);
        op['E'] = make_pair(1, 0);
        op['W'] = make_pair(-1, 0);

        if (n & 1) {
            cout << "NO" << endl;
            continue;
        }

        pair<int, int> maxx = findMax(s, n, op);
        if (maxx.first & 1 || maxx.second & 1) {
            cout << "NO" << endl;
            continue;
        }
        int targetX = maxx.first / 2;
        int targetY = maxx.second / 2;

        string re = "";
        bool flag = false;

        int x = 0, y = 0;
        if (targetX == 0 && targetY == 0) {
            bool flags[4] = {false, false, false, false};
            int idxs[4] = {-1, -1, -1, -1};

            for (int i = 0; i < n; i++) {
                if (s[i] == 'N') {
                    flags[0] = true;
                    idxs[0] = i;
                }
                else if (s[i] == 'S') {
                    flags[1] = true;
                    idxs[1] = i;
                }
                else if (s[i] == 'E') {
                    flags[2] = true;
                    idxs[2] = i;
                }
                else if (s[i] == 'W') {
                    flags[3] = true;
                    idxs[3] = i;
                }
                if (flags[0] && flags[1] && flags[2] && flags[3]) {
                    break;
                }
            }

            if (flags[0] && flags[1]) {
                re = string(n, 'R');
                re[idxs[0]] = 'H';
                re[idxs[1]] = 'H';
                if (n == 2) {
                    cout << "NO" << endl;
                } else {
                    cout << re << endl;
                }
                continue;
            }
            else if (flags[2] && flags[3]) {
                re = string(n, 'R');
                re[idxs[2]] = 'H';
                re[idxs[3]] = 'H';
                if (n == 2) {
                    cout << "NO" << endl;
                } else {
                    cout << re << endl;
                }
                continue;
            }
            else {
                cout << "NO" << endl;
                continue;
            }
        }
        bool flagR = false;
        bool flagH = false;
        for (int i = 0; i < n; i++) {
            if (flag) {
                re += string(n - i, 'R');
                flagR = true;
                break;
            }
            if (checkCloser(x, y, targetX, targetY, s[i])) {
                re += 'H';
                flagH = true;
                x += op[s[i]].first;
                y += op[s[i]].second;
            }
            else {
               re += 'R';
               flagR = true;
            }
            if (x == targetX && y == targetY) {
                flag = true;
            }
        }
        if (flagR && flagH) {
            cout << re << endl;
        }
        else {
            cout << "NO" << endl;
        }
    }
}

File: codeforces/Contests/Round#946/d.py
#!/usr/bin/python3

def checkCloser(x1, y1, targetX, targetY, c):
    if x1 < targetX and c == 'E':
        return True
    if x1 > targetX and c == 'W':
        return True
    if y1 < targetY and c == 'N':
        return True
    if y1 > targetY and c == 'S':
        return True
    return False

if __name__ == '__main__':
    # t = 1
    t = int(input())

    for _ in range(t):
        n = int(input())
        s = input()
        op = {'N': (0, 1), 'S': (0, -1), 'E': (1, 0), 'W': (-1, 0)}

        def Findmax():
            sett = set()
            maxX, maxY = 0, 0
            for c in s:
                sett.add(c)
                x, y = op[c]
                maxX += x
                maxY += y
            if len(sett) == 1 and n & 1:
                return None
            return maxX, maxY

        if not Findmax():
            print("NO")
            continue
        x, y = Findmax()
        if x & 1 or y & 1:
            print("NO")
            continue
        targetX, targetY = x // 2, y // 2
        re = ''
        flag = False

        x1, y1 = 0, 0
        used = [False] * n
        if targetX == 0 and targetY == 0:
            flags = [False, False, False, False]
            idxs = [-1, -1, -1, -1]
            for idx, c in enumerate(s):
                if c == 'N':
                    flags[0] = True
                    idxs[0] = idx
                if c == 'S':
                    flags[1] = True
                    idxs[1] = idx
                if c == 'E':
                    flags[2] = True
                    idxs[2] = idx
                if c == 'W':
                    flags[3] = True
                    idxs[3] = idx
                if flags[0] and flags[1] and flags[2] and flags[3]:
                    break
            if flags[0] and flags[1]:
                idx1 = min(idxs[0], idxs[1])
                idx2 = max(idxs[0], idxs[1])
                re = 'R' * idx1 + 'H' + 'R' * (idx2 - idx1 - 1) + 'H' + 'R' * (n - idx2 - 1)
                if n == 2:
                    print("NO")
                else:
                    print(re)
                continue
            elif flags[2] and flags[3]:
                idx1 = min(idxs[2], idxs[3])
                idx2 = max(idxs[2], idxs[3])
                re = 'R' * idx1 + 'H' + 'R' * (idx2 - idx1 - 1) + 'H' + 'R' * (n - idx2 - 1)
                if n == 2:
                    print("NO")
                else:
                    print(re)
                continue
            else:
                print("NO")
                continue

        for idx, c in enumerate(s):
            if flag:
                re += 'R' * (n - idx)
                break
            if checkCloser(x1, y1, targetX, targetY, c):
                re += 'H'
                x1 += op[c][0]
                y1 += op[c][1]
            else:
                re += 'R'
            if x1 == targetX and y1 == targetY:
                flag = True
        if len(set(re)) == 1:
            print("NO")
        else:
            print(re)
